proposal/color-4-new-spaces-js.d.ts.md

```diff
@@ -0,0 +1,864 @@
+
+export type RectangularColorSpace = Omit<KnownColorSpace, PolarColorSpace>;
+
+export type HueInterpolationMethod =
+  | 'decreasing'
+  | 'increasing'
+  | 'longer'
+  | 'shorter';
+```
+
+### New Color Functions
+
+```ts
+export class SassColor extends Value {
+```
+
+#### `space`
+
+Returns the name of [`internal`]'s space.
+
+[`internal`]: ../spec/js-api/value/color.d.ts.md#internal
+
+```ts
+get space(): KnownColorSpace;
+```
+
+#### `toSpace`
+
+Returns the result of [`color.to-space(internal, space)`] as a new SassColor.
+
+```ts
+toSpace(space: KnownColorSpace): SassColor;
+```
+
+[`color.to-space(internal, space)`]: ./color-4-new-spaces.md#colorto-space
+
+#### `isLegacy`
+
+Returns whether [`internal`] is in a [legacy color space] (`rgb`, `hsl`, or
+`hwb`).
+
+```ts
+get isLegacy(): boolean;
+```
+
+[legacy color space]: ./color-4-new-spaces.md#legacy-color
+
+#### `isInGamut`
+
+Returns the result of [`color.is-in-gamut(internal, space)`].
+
+```ts
+isInGamut(space?: KnownColorSpace): boolean;
+```
+
+[`color.is-in-gamut(internal, space)`]: ./color-4-new-spaces.md#coloris-in-gamut
+
+#### `toGamut`
+
+Returns the result of [`color.to-gamut(internal, space)`].
+
+```ts
+toGamut(space?: KnownColorSpace): SassColor;
+```
+
+[`color.to-gamut(internal, space)`]:  ./color-4-new-spaces.md#colorto-gamut-1
+
+#### `channelsOrNull`
+
+Returns an array of channel values for [`internal`], with [missing
+channels][missing components] converted to `null`.
+
+* Let `space` be the result of [`this.space`].
+
+* Let `components` be the list of channels in `space` and "alpha".
+
+* Let `channels` be an empty array.
+
+* For each `component` in `components`:
+
+  * Let `value` be the channel value in `color` with name of `component`.
+
+  * If `value` is `none`, let `value` be `null`.
+
+  * Append `value` to `channels`.
+
+* Return `channels`.
+
+```ts
+get channelsOrNull(): [
+  number | null,
+  number | null,
+  number | null,
+  number | null
+];
+```
+
+[missing components]: ./color-4-new-spaces.md#missing-components
+[`this.space`]: #space
+
+#### `channels`
+
+This algorithm returns an array of channel values for [`internal`], with
+[missing channels][missing components] converted to `0`.
+
+* Let `channels` be the result of [`this.channelsOrNull`].
+
+* For each `channel` in `channels`:
+
+  * If `value` equals `null`, let `value` be 0.
+
+  * Append `value` to `channels`.
+
+* Return `channels`.
+
+[`this.channelsOrNull`]: #channelsornull
+
+```ts
+get channels(): [number, number, number, number];
+```
+
+#### `channel`
+
+* Let `initialSpace` be the value of [`this.space()`].
+
+* Let `space` be `options.space` if it is defined, and the value of
+  `initialSpace` otherwise.
+
+* If `channel` is not "alpha" or a channel in `space`, throw an error.
+
+* If `space` is not equal to `initialSpace`, let `color` be the result
+  of [`this.toSpace(space)`]. Otherwise let `color` be `this`.
```

Comment:
I think you can just assign this unconditionally, since `toSpace()` should be a no-op for the same color space anyway.

Reply:
Updated in be40def4295de5511b6d2870742d877c68e982d9. I also updated this in `change`.